The Art of Component Composition - Leveraging Compound Components in React added (#21)
Zuehlke/software-engineering-corner: drafts/compound-components.md

---
title: The Art of Component Composition - Leveraging Compound Components in React
domain: software-engineering-corner.zuehlke.com
tags: reactjs, components, patterns

> cover: https://cdn.hashnode.com/res/hashnode/image/upload/v1709986147750/6D9yz4Rsi.png?auto=format

Reviewer: it looks great, but I am not a big fan of the blue ^^ maybe something darker would work better
<img width="1624" alt="image" src="https://github.com/Zuehlke/software-engineering-corner/assets/113620162/ce337ba8-4a8f-487c-bbe3-d2f453abbb1a">

Author: What do you think about the new version?

publishAs: gadu
hideFromHashnodeCommunity: false
saveAsDraft: true
---

As frontend applications get bigger and bigger, they have to cover more and more use cases and therefore their UI components need to get more flexible. This blog post describes how the requirements for a UI component increase throughout a typical project. We show what problems this entails and how they can be solved with the compound component pattern.

## Exploring the issue by the example of a dialog

> We have a React application that can manage files. One feature is to delete files, another is to duplicate them. For both actions, the action should be confirmed with a dialog. Since deleting is a destructive action, we also want to color the delete button red. In such a case, we want to implement a dialog component that we can use for both scenarios. We, therefore, need various parameters to configure the dialog accordingly. The following code shows how the call of such a component could look like:

Reviewer: ```suggestion
We have a React application that can manage files. One feature is to delete files, another is to duplicate them. For both actions, the action should be confirmed with a dialog. Since deleting is a destructive action, we also want to color the delete button red. In such a case, we want to implement a dialog component that we can use for both scenarios. We, therefore, need various parameters to configure the dialog accordingly. The Dialog Component could look like this:
```

```typescript
<Dialog
  title="Delete File"
  content="Do you really want to delete this file? You cannot undo this action."
  labelActionButton="Delete"
  isDestructiveAction
/>
```

The dialog now looks like this:

![Delete Dialog](https://cdn.hashnode.com/res/hashnode/image/upload/v1709470151176/qBBC62QhH.png?auto=format)

The next requirement is to add a renaming dialog, where you can enter the new file name. The content of the dialog does not only consist of text but also contains an input field. We must be able to pass this to the dialog.
Another requirement is to convert documents into images. The confirmation dialog should now have three buttons: "Cancel", "Convert to PNG" and "Convert to JPG". We now need to introduce a third button and make it configurable.
We are only at the beginning of our application and the number of parameters is already starting to get out of hand and there are still many requirements to come.

> If we don't want a huge, confusing dialog component that is full of configuration parameters and conditions, we need to take a different approach. One approach that is often chosen is to implement different subtypes of dialogs. Instead of one dialog, you have multiple dialogs for different use cases. One example could be a ConfirmDialog, a dialog that is exclusively used for confirmations of actions.

Reviewer: i am bad a punctuation, but for me it makes more sense to have a comma there :D 
```suggestion
If we don't want a huge, confusing dialog component that is full of configuration parameters and conditions, we need to take a different approach. One approach that is often chosen, is to implement different subtypes of dialogs. Instead of one dialog, you have multiple dialogs for different use cases. One example could be a ConfirmDialog, a dialog that is exclusively used for confirmations of actions.
```

Having different dialogs, however, carries the risk that they drift apart and get various looks and feels. There are often so many dialogs that some developers don't know all of them and start introducing new dialogs, even though there is already a dialog for a similar requirement.

This is where the compound component pattern comes to the rescue.

## The Compound Component Pattern

The compound component pattern is a component composition pattern. The aim is to have many small components instead of one large component, all of which require far fewer configuration parameters.
It always takes four steps to implement a compound component:

1. Create a Context
2. Create a parent component
3. Create child components
4. Add child components as properties to the parent components

All child components can be combined as desired and have a common context.
The shared context is what makes the pattern so powerful. Not many React developers know and use the pattern. However, the pattern is often used in component libraries as they often require a high level of flexibility. A well-known example is the Headless UI library from tailwind labs.

> In the following chapter we implement a dialog with the compound component pattern. The example will show how the different components are implemented and how the common context is created.

Reviewer: ```suggestion
Let's get started with implementing a dialog component using the compound component pattern. The example will show how the different components are implemented and how the common context is created.
```

## Implementing a Dialog with the Compound Component Pattern

Let's first see what a dialog looks like, what's common with all of them, and what has to be customizable.

![Parts of a dialog](https://cdn.hashnode.com/res/hashnode/image/upload/v1709464693100/g5q_83GTB.png?auto=format)

> - Title: The title has always the same font style but the text has to be customizable

Reviewer: make the name here bold *Title*, *Close Button*
I think this could help the reader

- Close Button: The close button has to be identical

> - Content: the content can be completely different

Reviewer: ```suggestion
- Content: The content can be completely different
```

  const { close } = useContext(DialogContext);

  return (
    <Button
      variant="primary"
      onClick={() => {
        onClick();
        close();
      }}
    >
      {label}
    </Button>
  );
};
```

4. After we define all the child components, we have to assign them to the parent component.

```typescript
// Dialog.tsx

// 4. Add child components as properties to the parent components
Dialog.Open = Open;
Dialog.Window = Window;
Dialog.Footer = Footer;
Dialog.CloseAction = CloseAction;
Dialog.CancelButton = CancelButton;
```

We defined our dialog and can start using it. The following example shows an example of the usage. It includes a button to open the dialog as well as the content of the dialog.

```typescript
// App.tsx

<Dialog>
  <Dialog.Open>
    <Button>Open Dialog</Button>
  </Dialog.Open>
  <Dialog.Window title="Delete File">
    Do you really want to delete this file? You cannot undo this action.
    <Dialog.Footer>
      <Dialog.CancelButton />
      <Dialog.ActionButton
        label="Delete"
        onClick={() => delete()}
      />
    </Dialog.Footer>
  </Dialog.Window>
</Dialog>
```

> You are very flexible with the configuration of the dialog. You don't need to close or open the dialog yourself, it is handled for you. It is easy to add new buttons with different actions or to add your own components to the dialog. In the following GitHub repository, you will find an additional example with an input field: https://github.com/gabduss/customizable-modal

Reviewer: Would it be possible to do a StackBlitz example as well?
